commons/utility/crypto.ts

```
@@ -64,13 +67,17 @@ export const asymmetricEncrypt = (
 	senderPrivateKey: EncodedUInt8Array,
 	senderPublicKey: EncodedUInt8Array,
 	receiverPublicKey: EncodedUInt8Array,
-	message: UnencryptedPayload
+	message: UnencryptedPayload,
+	passedNonce: Uint8Array
 ): AsymmetricEncrypted => {
 	const privKeyUInt8Array = util.decodeBase64(senderPrivateKey);
 	const pubKeyUInt8Array = util.decodeBase64(receiverPublicKey);
 	const msgParamsUInt8Array = util.decodeUTF8(message);
 
-	const nonce = nacl.randomBytes(nacl.box.nonceLength);
+	let nonce = passedNonce;
+	if (!nonce) {
+		nonce = nacl.randomBytes(nacl.box.nonceLength);
+	}
```

I think we need to return an error if nonce is not passed, say we forgot to pass the nonce when encrypting multiple files together, in this case, we will generate different nonces for every encryption?

---

What is the use case when you need a "passedNonce"?

---

from what I understand, passed nonce is the nonce that is passed by the `asymmetricEncrypt` implementor, that way we can encrypt multiple data using a single nonce, rather than generating a new one each time encryption is called

---

okay then, if that's the case, @altairrostam  please change the function type to
```typescript
export const asymmetricEncrypt = (
	senderPrivateKey: EncodedUInt8Array,
	senderPublicKey: EncodedUInt8Array,
	receiverPublicKey: EncodedUInt8Array,
	message: UnencryptedPayload,
	passedNonce?: Uint8Array
): AsymmetricEncrypted => {
```

Question mark in here signifies that `passedNonce` maybe null

---

but it shouldn't be possible to be null right? because we always need the nonce 🤔

---

```typescript
if (!nonce) {
    nonce = nacl.randomBytes(nacl.box.nonceLength);
}
```

this block will generate the nonce if null

---

Let's play safe with typings by annotating variables and functions correctly

---

The reason why there is passedNonce is because in the virtual accounts creation, the same nonce has to be used throughout the data.

So far in the codebase, the usage for the asymmetricEncrypt occurred twice; once in the storeSecretPhrase function and the other here in the bulk virtual account creation.

In the storeSecretPhrase function, it happened only once so the generation of nonce is made inside the asymmetricEncrypt.

---

ok, please adjust the typings, and will approve the PR afterward

---

I see, all good then, thanks

---

@linerocks ok will do